File: WorkIt/app/src/main/java/prot3ct/workit/data/remote/base/TaskDataContract.java
package prot3ct.workit.data.remote.base;

import java.util.List;

import io.reactivex.Observable;
import prot3ct.workit.view_models.AssignedTasksListViewModel;
import prot3ct.workit.view_models.AvailableTasksListViewModel;
import prot3ct.workit.view_models.CompletedTasksListViewModel;
import prot3ct.workit.view_models.IsUserAssignableToTaskViewModel;
import prot3ct.workit.view_models.MyTasksListViewModel;
import prot3ct.workit.view_models.TaskDetailViewModel;

public interface TaskDataContract {
    Observable<Boolean> createTask(String title, String startDate, String length,
                                   String description, String city, String address, String reward);

    Observable<Boolean> updateTask(int taskId, String title, String startDate, String length,
                                   String description, String city, String address, String reward);

    Observable<TaskDetailViewModel> getTaskDetails(int taskId);

    Observable<Boolean> deleteTask(int taskId);

    Observable<IsUserAssignableToTaskViewModel> canAssignToTask(int taskId);

    Observable<List<AssignedTasksListViewModel>> getAssignedTasks();

    Observable<List<AvailableTasksListViewModel>> getAvailableTasks(int page, String search);

    Observable<List<MyTasksListViewModel>> getMyTasks();

    Observable<List<CompletedTasksListViewModel>> getCompletedTasks();

    Observable<Boolean> removeAssignedUser(int taskId);
}

File: WorkIt/app/src/main/java/prot3ct/workit/data/remote/base/TaskDataContract.kt
package prot3ct.workit.data.remote.base

import io.reactivex.Observable
import prot3ct.workit.view_models.TaskDetailViewModel
import prot3ct.workit.view_models.IsUserAssignableToTaskViewModel
import prot3ct.workit.view_models.AssignedTasksListViewModel
import prot3ct.workit.view_models.AvailableTasksListViewModel
import prot3ct.workit.view_models.MyTasksListViewModel
import prot3ct.workit.view_models.CompletedTasksListViewModel

interface TaskDataContract {
    fun createTask(
        title: String, startDate: String, length: String,
        description: String, city: String, address: String, reward: String
    ): Observable<Boolean>

    fun updateTask(
        taskId: Int, title: String, startDate: String, length: String,
        description: String, city: String, address: String, reward: String
    ): Observable<Boolean>

    fun getTaskDetails(taskId: Int): Observable<TaskDetailViewModel>

    fun deleteTask(taskId: Int): Observable<Boolean>

    fun canAssignToTask(taskId: Int): Observable<IsUserAssignableToTaskViewModel>

    val assignedTasks: Observable<List<AssignedTasksListViewModel>>

    fun getAvailableTasks(
        page: Int,
        search: String
    ): Observable<List<AvailableTasksListViewModel>>

    val myTasks: Observable<List<MyTasksListViewModel>>

    val completedTasks: Observable<List<CompletedTasksListViewModel>>

    fun removeAssignedUser(taskId: Int): Observable<Boolean>
}